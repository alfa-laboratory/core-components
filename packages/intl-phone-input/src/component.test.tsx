import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { IntlPhoneInput } from './index';

describe('IntlPhoneInput', () => {
    const testId = 'intl-phone-input';

    it('should match snapshot', () => {
        const { container } = render(
            <IntlPhoneInput value='+7 111 111 11 11' onChange={jest.fn()} />,
        );

        expect(container).toMatchSnapshot();
    });

    it('should unmount without errors', () => {
        const { unmount } = render(
            <IntlPhoneInput value='+7 111 111 11 11' onChange={jest.fn()} />,
        );

        expect(unmount).not.toThrowError();
    });

    it('should call `onChange` callback after input was changed with dial code of country without priority', () => {
        const onChange = jest.fn();
        const { getByTestId } = render(
            <IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />,
        );

        const input = getByTestId(testId);

        fireEvent.change(input, { target: { value: '+54' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+54');
    });

    it('should call `onChange` callback after input was changed with dial code of country from NANP', () => {
        const onChange = jest.fn();
        const { getByTestId } = render(
            <IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />,
        );

        const input = getByTestId(testId);
        fireEvent.change(input, { target: { value: '+1868' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+1868');
    });

    it('should call `onChange` callback after input was changed with whole russian number', () => {
        const onChange = jest.fn();
        const { getByTestId } = render(
            <IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />,
        );

        const input = getByTestId(testId);
        fireEvent.change(input, { target: { value: '+74957888878' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('+74957888878');
    });

    it('should have default country flag icon', () => {
        const { container } = render(<IntlPhoneInput value='' onChange={jest.fn()} />);

        const flagComponent = container.querySelector('.flagIcon');

        expect(flagComponent).toHaveClass('ru');
    });

    it('should have passed country flag icon', () => {
        const { container } = render(
            <IntlPhoneInput value='' onChange={jest.fn()} defaultCountryIso2='az' />,
        );

        const flagComponent = container.querySelector('.flagIcon');

        expect(flagComponent).toHaveClass('az');
    });

    it('should set new country flag icon from props', async () => {
        const { container } = render(<IntlPhoneInput value='+61' onChange={jest.fn()} />);

        await waitFor(() => expect(container.querySelector('.flagIcon')).toHaveClass('au'));
    });

    it('should call `onChange` callback after select was changed', async () => {
        const onChange = jest.fn();

        const { container, getAllByRole } = render(<IntlPhoneInput value='' onChange={onChange} />);

        const flagComponent = container.querySelector('.flagIcon');

        fireEvent.click(flagComponent as HTMLSpanElement);

        const option = getAllByRole('option')[0];

        fireEvent.click(option);

        expect(onChange).toHaveBeenCalled();
    });

    it('should focus on input after select was changed', () => {
        const onChange = jest.fn();

        const { container, getAllByRole, getByTestId } = render(
            <IntlPhoneInput value='' onChange={onChange} dataTestId={testId} />,
        );

        const input = getByTestId(testId);

        const flagComponent = container.querySelector('.flagIcon');

        fireEvent.click(flagComponent as HTMLSpanElement);

        const option = getAllByRole('option')[0];

        fireEvent.click(option);

        expect(document.activeElement).toBe(input);
    });
});
