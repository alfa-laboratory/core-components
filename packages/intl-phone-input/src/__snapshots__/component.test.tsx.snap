// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlPhoneInput should match snapshot 1`] = `
<div>
  <div
    class="component m filled hasLeftAddons formControl m"
  >
    <div
      class="inner"
    >
      <div
        class="addons leftAddons addons"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="downshift-1-menu"
          class="component component"
          role="combobox"
        >
          <div
            class="component m"
          >
            <div
              aria-controls="downshift-1-menu"
              aria-labelledby="downshift-1-label"
              class="inner"
              id="downshift-0-input"
              tabindex="0"
            >
              <span>
                <span
                  class="flagIcon m ru"
                />
              </span>
              <span
                class="arrow"
              />
            </div>
          </div>
          <div
            aria-labelledby="downshift-1-label"
            class="listWrapper"
            id="downshift-1-menu"
            role="listbox"
          />
        </div>
      </div>
      <div
        class="inputWrapper"
      >
        <div
          class="input"
        >
          <input
            class="input"
            type="tel"
            value="+7 111 111 11 11"
          />
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
